### PR TITLE
Ports dynamic gamemode from /tg/ (which was ported from /vg/)

### PR DESCRIPTION
**Pull request by ghost**
https://github.com/tgstation/tgstation/pull/44639

@Manly-Dwarf, I respect your argument that you make on discord a while back, but nonetheless I support this mode.
> I think it's important to have some patience here. We already have two gamemodes in the works, apparently pillar men is 80% done  

It'll probably never be done for a while tbh, the remaining 20% is creativity, which I lack, and nobody is willing to help me with.

> plus goonops, plus the security overhaul, plus more

I haven't seen a single line of code for Goonops, plus isn't the security overhaul on hold?

> Dynamic is a gamemode, which means it needs to go into rotation, and there's no point having vampire, traitorvamp and traitor all in rotation at the same time along with it

I'm not going to be making Vampire a dynamic ruleset, because i'm not a sadist. There IS a point though -- different "round styles". A traitor round would just be, well, traitors, but dynamic could be traitors, wizards, and a hivemind jerking each other off.

> Some discussion would be needed.

Well here's the place to discuss!

>If you want my two cents, I like the idea, but I'd probably chip in my own system. For instance, an extremely rare chance for one of the people selected to roll antagonist gets to choose what antagonist he wants to be? Might result in some unknown antagonists like Hivemind, Devil, and others of dubious quality being allowed to show up. Could enable some fun things, don't you think?

I like this idea.

> If I wanted to merge something like Dynamic, I'd do it along with the Bounty Hunters mode I discussed with you

Bounty hunters is long from done though.

> That'd make the system simple. Lowpop, you get Traitor. Past 10 or 15, you either get Bounties, Dynamic, or the team antag modes.

What?

### CODE REVIEW

**Comment by ghost** [comment]
note: do not squash merge

**Comment by Manly-Dwarf** [comment]
This sounds like it could end terribly, especially if Wizard or conversion antagonists show up at all ever with other antagonists previously active, or made active at the same time.
It'd be worth giving a shot, no guarantees.

**Comment by Manly-Dwarf** [comment]
Also goonops is already coded, there's just one last bug that I'm too lazy to fix and push the PR.

**Comment by ghost** [comment]
some notes to anyone reading this: only one type of round-ending antagonist can roll at the same time, if at all.

revs, cult, and nukies are considered round-ending

**Comment by Manly-Dwarf** [comment]
but not wiz with his badmin gauntlet?
Gotta be careful around that one, too. Badmin Gauntlet needs extra precautions to be added to ensure you can't buy it past roundstart, or if there's other antagonists active.

**Comment by ghost** [comment]
oh, good idea. i'll just uh, blacklist badmin gauntlet during dynamic

**Comment by Manly-Dwarf** [comment]
lazy ass

File: hippiestation/code/modules/antagonists/wizard/equipment/spellbook.dm [hunk 1]
```diff
@@ -138,7 +138,7 @@
 /datum/spellbook_entry/item/badmin_gauntlet/IsAvailible()
 	if(!..())
 		return FALSE
-	return (SSticker.mode.name != "ragin' mages") && !GLOB.gauntlet_equipped
+	return !istype(SSticker.mode, /datum/game_mode/wizard/raginmages) && !istype(SSticker.mode, /datum/game_mode/dynamic) && !GLOB.gauntlet_equipped
```
Comment (Manly-Dwarf) [inline hippiestation/code/modules/antagonists/wizard/equipment/spellbook.dm:141]: lazy ass

Comment (ghost) [reply hippiestation/code/modules/antagonists/wizard/equipment/spellbook.dm:141]: lmao

**Comment by TheChosenEvilOne** [comment]
well then, have fun with this.

**Comment by ghost** [comment]
@TheChosenEvilOne what do you mean?

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 2]
```diff
@@ -0,0 +1,751 @@
...
+/datum/game_mode/dynamic/proc/generate_threat()
+	var/relative_threat = LORENTZ_DISTRIBUTION(GLOB.dynamic_curve_centre, GLOB.dynamic_curve_width)
+	threat_level = round(lorentz_to_threat(relative_threat), 0.1)
+
+	peaceful_percentage = round(LORENTZ_CUMULATIVE_DISTRIBUTION(relative_threat, GLOB.dynamic_curve_centre, GLOB.dynamic_curve_width), 0.01)*100
+
+	threat = threat_level
+
+/datum/game_mode/dynamic/can_start()
+	/* Disabled for now, had some changes that need to be tested and this might interfere with that.
```
Comment (ghost) [inline code/game/gamemodes/dynamic/dynamic.dm:268]: Hey @TheChosenEvilOne is there a reason this is still commented out?

**Comment by Manly-Dwarf** [comment]
find a better way to get the badmin gauntlet working in dynamic. maybe if it selects a wizard, give him like a three minute grace period to buy the gauntlet and if he fails in that time spawn antags? I dunno.

**Comment by Manly-Dwarf** [comment]
also disable blob & blood brothers, add hivemind & abductors

**Comment by ghost** [comment]
@Manly-Dwarf should abductors be roundstart? i know hiveminds should be, but abductors?

**Comment by Manly-Dwarf** [comment]
ye why not

**Comment by Manly-Dwarf** [comment]
they can cause a lot of chaos if they live for too long

**Comment by ghost** [comment]
@Manly-Dwarf should the gauntlet grace period check to see if the gauntlet EXISTS or is equipped? 

I can see wizards simply buying the badmin gauntlet and then refunding it after the 3-minute mark if it's the former.

**Comment by Manly-Dwarf** [comment]
exists, if the wizard refunds it then instantly create traitors and block his purchase again or somethin I dunno

**Comment by ghost** [comment]
actually, this is impossible without major refactoring :(

**Comment by ghost** [comment]
also, blob isn't roundstart, @Manly-Dwarf

File: code/game/gamemodes/dynamic/dynamic.dm (hunk 2 above)
Comment (TheChosenEvilOne) [reply code/game/gamemodes/dynamic/dynamic.dm:268]: Simply, it was merged before I could uncomment it.

**Comment by ghost** [comment]
so how will this be handled if testmerges are borken?

File: code/game/gamemodes/dynamic/dynamic.dm (hunk 2 above)
Comment (ghost) [reply code/game/gamemodes/dynamic/dynamic.dm:268]: Oh, ok.

**Comment by SireTurret** [comment]
boomer gamemode

**Comment by Manly-Dwarf** [comment]
This will probably just be merged, tested out, if things go wrong we'll patch them out down the line. It's unlikely that this won't work any way.

**Review by Manly-Dwarf — APPROVED** [review]
Should be fine from what I see. If Yoyo doesn't leave a review soon I'll merge it and we'll start testing it.

**Comment by steamp0rt** [comment]
YEET
image

**Comment by Manly-Dwarf** [comment]
Here goes nothing.

**Comment by Manly-Dwarf** [comment]
From what I can see so far, I think I have a couple complaints? Although it might be thanks to the randomness of the mode:
The spending is not nearly aggressive enough, it leaves 20 points open roundstart and sits on them like a hoarder.
Takes a bit too long for it to spawn midrounds too.
All the rounds I've tested it so far, it's only spawned two antagonists, both the same type. No hivemind ling or hivemind traitor, just hivemind or traitor.

**Comment by Manly-Dwarf** [comment]
it must be WILDER steamp0rt

**Comment by ghost** [comment]
@Manly-Dwarf https://github.com/HippieStation/HippieStation/pull/11921

**Pull request merged**